src/rust/ide/src/controller/project.rs

```diff
@@ -10,44 +10,145 @@ use crate::controller::Visualization;
+        let ctx = ctx.upgrade().ok_or_else(|| VisualizationDoesNotExistAnymore(context_id))?;
+        ctx.dispatch_visualization_update(visualization_id,data)
+    }
+
+    /// Registers a new ExecutionContext. It will be eligible for receiving future updates routed
+    /// through `dispatch_visualization_update`.
+    pub fn insert(&self, context:&Rc<ExecutionContext>) {
+        self.0.borrow_mut().insert(context.id(),Rc::downgrade(&context));
+    }
+}
+
 
 // ==========================
 // === Project Controller ===
 // ==========================
 
-type ModulePath = controller::module::Path;
-
 /// Project controller's state.
 #[allow(missing_docs)]
-#[derive(Debug)]
+#[derive(Clone,CloneRef,Debug)]
 pub struct Handle {
+    pub project_name        : Rc<String>,
     pub language_server_rpc : Rc<language_server::Connection>,
     pub visualization       : Visualization,
     pub language_server_bin : Rc<binary::Connection>,
     pub module_registry     : Rc<model::registry::Registry<ModulePath,model::synchronized::Module>>,
+    pub execution_contexts  : Rc<ExecutionContextsRegistry>,
     pub parser              : Parser,
     pub logger              : Logger,
 }
 
 impl Handle {
     /// Create a new project controller.
     pub fn new
-    ( parent                 : &Logger
-    , language_server_client : language_server::Connection
-    , language_server_binary : binary::Connection
+    ( parent                     : &Logger
+    , language_server_client     : language_server::Connection
+    , mut language_server_binary : binary::Connection
+    , project_name               : impl Str
     ) -> Self {
-        let module_registry         = default();
-        let parser                  = Parser::new_or_panic();
+        let logger = parent.sub("Project Controller");
+        info!(logger,"Creating a project controller for project {project_name.as_ref()}");
+        let binary_protocol_events  = language_server_binary.event_stream();
+        let embedded_visualizations = default();
         let language_server_rpc     = Rc::new(language_server_client);
         let language_server_bin     = Rc::new(language_server_binary);
-        let logger                  = parent.sub("Project Controller");
-        let embedded_visualizations = default();
         let language_server         = language_server_rpc.clone();
         let visualization           = Visualization::new(language_server,embedded_visualizations);
-        Handle {module_registry,parser,language_server_rpc,language_server_bin,logger,visualization}
+
+        let ret = Handle {
+            project_name        : Rc::new(project_name.into()),
+            module_registry     : default(),
+            execution_contexts  : default(),
+            parser              : Parser::new_or_panic(),
+            language_server_rpc,
+            language_server_bin,
+            logger,
+            visualization,
+        };
```

Comment:
I would opt for having all fields defined on lets, or at least have all fields taken from lets in one line, to avoid vertical noise

Reply:
Done